ml/train/losses/losses.go

```diff
@@ -398,3 +402,32 @@ func MakeHuberLoss(delta float64) LossFn {
 		return loss
 	}
 }
+
+// CosineSimilarityLoss computes the cosine similarity loss between two predictions.
+//
+// It assumes two predictions are provided which should be tensors of rank 2 with the same dimensions (N, D).
+// It also assumes there is one label tensor with rank 2 with dimensions (N, 1) that contains the target cosine similarity.
+// The loss will then calculate the residuals between the cosine similarity of the two prediction tensors row-wise
+// and the target cosine similarity, and return the mean of the squared residuals.
+func CosineSimilarityLoss(labels, predictions []*context.Node) *context.Node {
+	predictionsLeft := predictions[0]
+	predictionsRight := predictions[1]
+	scores := labels[0]
+
+	if predictionsLeft.Shape().Rank() != 2 {
+		panic(fmt.Errorf("expected rank 2, got %d", predictionsLeft.Shape().Rank()))
```

Comment:
I've been using `exceptions.Panicf` that does the same, except it uses the `errors` package, which includes a stacktrace -- helpful for debugging errors.

Optional, but for extra help, maybe print out all the shape of the offending value (it often helps for one to realize what they were doing)

exceptions.Panicf("expected rank 2, got %d (shape=%s)", predictionsLeft.Shape().Rank(), predictionsLeft.Shape())

Same below.

Reply:
Actually, `exceptions` is also imported as ".". It only exposes the `Panicf()` function. 

So simply: `Panicf("...")`